File: typescript/src/era.ts
import { ymd, type Ymd } from "./dates.js";

/** 元号（和暦）の定義。改元日 (y/m/d) で区切る。Dart / Java 版と同一テーブル。 */
export interface Era {
  name: string;
  abbr: string;
  y: number;
  m: number;
  d: number;
}

/** 組込みの元号テーブル（新しい順）。wareki フォーマッタと eraOf が共有する。 */
export const ERAS: Era[] = [
  { name: "令和", abbr: "R", y: 2019, m: 5, d: 1 },
  { name: "平成", abbr: "H", y: 1989, m: 1, d: 8 },
  { name: "昭和", abbr: "S", y: 1926, m: 12, d: 25 },
  { name: "大正", abbr: "T", y: 1912, m: 7, d: 30 },
  { name: "明治", abbr: "M", y: 1868, m: 10, d: 23 },
];

/** eraOf の結果。元号名・略記・和暦年（元年 = 1）。 */
export interface EraDate {
  name: string;
  abbr: string;
  year: number;
}

function cmp(a: Ymd, e: Era): number {
  return a.y - e.y || a.m - e.m || a.d - e.d;
}

/** ymd から元号を算出する（明治より前は null）。 */
export function eraOfYmd(p: Ymd): EraDate | null {
  const era = ERAS.find((e) => cmp(p, e) >= 0);
  if (!era) return null;
  return { name: era.name, abbr: era.abbr, year: p.y - era.y + 1 };
}

/** 日付の元号を算出する（明治より前は null）。Dart / Java 版と同名・同出力。 */
export function eraOf(date: string | Date): EraDate | null {
  return eraOfYmd(ymd(date));
}

File: java/src/main/java/io/hatake/core/Era.java
package io.hatake.core;

import java.util.List;

/**
 * 元号（和暦）の算出。組込みの元号テーブルを {@code wareki} フォーマッタと共有する。
 * Dart / TypeScript 版と同一テーブル・同出力。
 */
public final class Era {
    private Era() {
    }

    /** 元号の定義（改元日 y/m/d で区切る）。 */
    public record EraDef(String name, String abbr, int y, int m, int d) {
    }

    /** eraOf の結果。元号名・略記・和暦年（元年 = 1）。 */
    public record EraDate(String name, String abbr, int year) {
    }

    /** 組込みの元号テーブル（新しい順）。 */
    public static final List<EraDef> ERAS = List.of(
            new EraDef("令和", "R", 2019, 5, 1),
            new EraDef("平成", "H", 1989, 1, 8),
            new EraDef("昭和", "S", 1926, 12, 25),
            new EraDef("大正", "T", 1912, 7, 30),
            new EraDef("明治", "M", 1868, 10, 23));

    private static int cmp(int y, int m, int d, EraDef e) {
        if (y != e.y()) {
            return y - e.y();
        }
        if (m != e.m()) {
            return m - e.m();
        }
        return d - e.d();
    }

    /** ymd から元号を算出する（明治より前は null）。 */
    public static EraDate eraOfYmd(int y, int m, int d) {
        for (EraDef e : ERAS) {
            if (cmp(y, m, d, e) >= 0) {
                return new EraDate(e.name(), e.abbr(), y - e.y() + 1);
            }
        }
        return null;
    }

    /** 日付（yyyy-MM-dd）の元号を算出する（明治より前は null）。 */
    public static EraDate eraOf(String date) {
        var dt = Dates.parse(date);
        return eraOfYmd(dt.getYear(), dt.getMonthValue(), dt.getDayOfMonth());
    }
}
